File: Data Structures & Algorithms/search-2d-matrix/submission-1.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size();
        int col = matrix[0].size();
        int low = 0;
        int high = (row * col) - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            int r = mid / col;
            int c = mid % col;
            int val = matrix[r][c];

            if (val < target) {
                low = mid + 1;
            }
            else if (val > target) {
                high = mid - 1;
            }
            else {
                return true;
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low, high = 1, max(piles)

        while low <= high:
            mid = low +(high - low) // 2
            totalHours = 0
            for pile in piles:
                totalHours += ((pile + mid - 1) // mid)
            if totalHours > h:
                low = mid + 1
            else:
                high = mid - 1
        
        return low